File: leetcode-in-python3/solution/5_longest_palindromic_substring.py

class Solution:
    def longestPalindrome(self, s: str) -> str:
        ans = ""
        for index, _ in enumerate(s):
            s1 = self._longestPalindrome(s, index, index)
            s2 = self._longestPalindrome(s, index, index + 1)
            ss = s2 if len(s1) <= len(s2) else s1
            if len(ans ) < len(ss):
                ans = ss
        return ans

    def _longestPalindrome(self, s: str, l: int, r: int):
        while 0 <= l and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        return s[l+1:r]


import unittest


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.input = ["babad", "cbbd", "a", "", "ab", "aa"]
        self.answer = ["bab", "bb", "a", "", "a", "aa"]
        self.s = Solution()

    def test_solution(self):
        for i, s in enumerate(self.input):
            ans = self.s.longestPalindrome(s)
            self.assertEqual(ans , self.answer[i],ans )


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_5_longest_palindromic_substring.rs
pub struct Solution;

impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let s = s.as_bytes();
        let mut ans = "";
        for (i, _) in s.iter().enumerate() {
            let s1 = Solution::_longest_palindrome(s, i, i);
            let s2 = Solution::_longest_palindrome(s, i, i + 1);
            let ss = if s1.len() < s2.len() { s2 } else { s1 };
            if ans.len() < ss.len() {
                ans = ss;
            }
        }
        ans.to_string()
    }

    fn _longest_palindrome(s: &[u8], l: usize, r: usize) -> &str {
        let (mut l, mut r) = (l, r);
        // 0 <= usize always true, It means 0 <= l always true
        while r < s.len() && s[l] == s[r] {
            if l == 0 {
                return std::str::from_utf8(&s[l..r + 1]).unwrap();
            }
            l -= 1;
            r += 1;
        }
        std::str::from_utf8(&s[l + 1..r]).unwrap()
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            ("babad".to_string(), "bab"),
            ("cbbd".to_string(), "bb"),
            ("a".to_string(), "a"),
            ("".to_string(), ""),
            ("ab".to_string(), "a"),
            ("aa".to_string(), "aa"),
        ];
        for (item, answer) in test_case {
            let ans = Solution::longest_palindrome(item);
            assert_eq!(answer.to_string(), ans);
        }
    }
}
